resources/Microsoft.Windows.Developer/Microsoft.Windows.Developer.psm1

```diff
@@ -800,6 +805,64 @@ class WindowsCapability {
         }
     }
 }
+
+[DSCResource()]
+class AdvancedNetworkSharingSetting {
+    [DscProperty(Key, Mandatory)]
+    [AdvancedNetworkSharingSettingName]$Name
+
+    [DscProperty()]
+    [string[]]$Profiles = @()
+
+    [DscProperty(NotConfigurable)]
+    [string[]]$EnabledProfiles
+
+    hidden [string] $NetworkDiscoveryGroup = '@FirewallAPI.dll,-32752'
+    hidden [string] $FileAndPrinterSharingGroup = '@FirewallAPI.dll,-28502'
+
+    [AdvancedNetworkSharingSetting] Get() {
+
+        $currentState = [AdvancedNetworkSharingSetting]::new()
+        $currentState.Name = $this.Name
+        $currentState.Profiles = $this.Profiles
+
+        if ($this.Name -eq [AdvancedNetworkSharingSettingName]::NetworkDiscovery) {
+            $group = $this.NetworkDiscoveryGroup
+        } else {
+            $group = $this.FileAndPrinterSharingGroup
+        }
+
+        $this.EnabledProfiles = Get-NetFirewallRule -Group $group | Where-Object { $_.Enabled -eq 'True' } | Select-Object -Unique -CaseInsensitive -ExpandProperty Profile
+        $currentState.EnabledProfiles = $this.EnabledProfiles
+
+        return $currentState
+    }
+
+    [bool] Test() {
+        $currentState = $this.Get()
+
+        # Compare-object is case insensitive by default and does not take null arguments
+        $difference = Compare-Object -ReferenceObject @( $this.Profiles | Select-Object) -DifferenceObject @( $currentState.EnabledProfiles | Select-Object)
+        return -not $difference
+    }
+
+    [void] Set() {
+        if (!$this.Test()) {
+            if ($this.Name -eq [AdvancedNetworkSharingSettingName]::NetworkDiscovery) {
+                $group = $this.NetworkDiscoveryGroup
+            } else {
+                $group = $this.FileAndPrinterSharingGroup
+            }
+
+            $firewallGroups = Get-NetFirewallRule -Group $group
+            #Enable
+            $firewallGroups | Where-Object { ($_.Enabled -eq 'False') -and ($this.Profiles -Contains $_.Profile ) } | Set-NetFirewallRule -Enabled True
```

Comment:
I ran into a corner case on "File and printer sharing" in which a "private" rule did not exist, causing this logic to not work to enable the setting. I toggled the setting via the UX and then this logic worked as the UX had created the firewall rules.. What is the correct way to enable this scenario via PS without the UX? [ Searching online](https://www.c-sharpcorner.com/article/how-to-enable-or-disable-file-and-printer-sharing-in-windows-102/) makes me wonder if Set-NetFirewallRule on the group directly would do the right thing, or if we need to manually create the firewall rule if it does not initially exist.

Reply:
I have validated with testing on default firewalls rules that "Set-NetFirewallRule" targeting specific profiles produces the intended result.